[flang] use TBAA Forest in TBAABuilder

tblah opened the pull request:
See previous reviews at https://github.com/llvm/llvm-project/pull/68414 and https://github.com/llvm/llvm-project/pull/68317

This is important to ensure that tags end up in the same trees that were
created in the FIR TBAA pass. If they are in different trees then
everything in one tree will be assumed to MayAlias with everything in the
other tree. This leads to poor performance.

After this change, the trees generated by TBAABuilder are separated per-function. This could lead to less precise alias information after in-lining, but no serious performance regressions have been found so far.

@vzakhari requested that the old (not-per-function) trees are
maintained so I left the old test intact.

tblah committed:
[flang][FIR] add FirAliasAnalysisOpInterface

This interface allows (HL)FIR passes to add TBAA information to fir.load
and fir.store. If present, these TBAA tags take precedence over those
added during CodeGen.

We can't reuse mlir::LLVMIR::AliasAnalysisOpInterface because that uses
the mlir::LLVMIR namespace so it tries to define methods for fir
operations in the wrong namespace. But I did re-use the tbaa tag type to
minimise boilerplate code.

The new builders are to preserve the old interface without the tbaa tag.

tblah committed:
[flang][FIR] add FIR TBAA pass

See RFC at
https://discourse.llvm.org/t/rfc-propagate-fir-alias-analysis-information-using-tbaa/73755

This pass adds TBAA tags to all accesses to non-pointer/target dummy
arguments. These TBAA tags tell LLVM that these accesses cannot alias:
allowing better dead code elimination, hoisting out of loops, and
vectorization.

Each function has its own TBAA tree so that accesses between funtions
MayAlias after inlining.

I also included code for adding tags for local allocations and for
global variables. Enabling all three kinds of tag is known to produce a
miscompile and so these are disabled by default. But it isn't much
code and I thought it could be interesting to play with these later if
one is looking at a benchmark which looks like it would benefit from
more alias information. I'm open to removing this code too.

TBAA tags are also added separately by TBAABuilder during CodeGen.
TBAABuilder has to run during CodeGen because it adds tags to box
accesses, many of which are implicit in FIR. This pass cannot (easily)
run in CodeGen because fir::AliasAnalysis has difficulty tracing values
between blocks, and by the time CodeGen runs, structured control flow
has already been lowered.

Coming in follow up patches
  - Change CodeGen/TBAABuilder to use TBAAForest to add tags within the
    same per-function trees as are used here (delayed to a later patch
    to make it easier to revert)
  - Command line argument processing to actually enable the pass

tblah committed:
[flang] use TBAAForest in TBAABuilder

This is important to ensure that tags end up in the same trees that were
created in the FIR TBAA pass. If they are in different trees then
everything in one tree will be assumed to MayAlias with everything in the
other tree. This leads to poor performance.

@vzakhari requested that the old (not-per-function) trees are
maintained so I left the old test intact.

tblah committed:
[flang][FIR] Add dependency from FIRDialect to LLVMDialect

For TBBAA attribute types.

vzakhari commented on flang/lib/Optimizer/CodeGen/TBAABuilder.cpp:
```diff
@@ -39,26 +49,10 @@ static llvm::cl::opt<unsigned>
 TBAABuilder::TBAABuilder(MLIRContext *context, bool applyTBAA)
-    : enableTBAA(applyTBAA && !disableTBAA) {
+    : enableTBAA(applyTBAA && !disableTBAA),
+      trees(/*separatePerFunction=*/perFunctionTBAATrees) {
```
Do I understand it correctly that even without running the FIR TBAA attachment pass we will use per-function trees by default after this change?  I wonder if we can still stick to a single global tree for those functions that have not been affected by the FIR TBAA attachment pass?

tblah replied:
Yes we will. Although, my intention is to enable my FIR pass whenever we are using TBAA so I don't anticipate TBAABuilder will often run without the FIR alias tags pass..

I am yet to see evidence of significant slowdowns due to using per-function TBAA trees. The FIR pass applies tags to every function which has dummy arguments, so I suspect keeping the global tree for unaffected functions which slow down due to per-function trees is a very uncommon case.

vzakhari replied:
Thanks for explaining.  My only concern is a "clean" switch back to the previous state of things (e.g. to triage new failures, if any).  With this change, we will have to specify two options to switch to the TBAA that we currently have, i.e. we will have to disable the new FIR TBAA pass, and also disable the per-function behavior for the codegen.  It is not a big deal, I just wanted to confirm that this is the case.

vzakhari replied:
I think so, but I am okay with having the two flags as well.

tblah replied:
Done. See https://github.com/llvm/llvm-project/pull/68437/commits/56ed3a6e25c34542a47d52afc7cf1aeaa2b58271 and https://github.com/llvm/llvm-project/pull/68595/commits/9168aa0f6722d576267747e3ec3f8178e6df7f85

vzakhari replied:
Thank you! It looks good to me.

tblah committed:
[flang] Rename FirAliasAnalysisOpInterface to FirAliasTagOpInterface

tblah committed:
[flang] Rename FirAliasAnalysisOpInterface to FirAliasTagOpInterface

tblah committed:
Fix typos

tblah committed:
Rename pass to fir-add-alias-tags

tblah committed:
Get function argument name from fir::DeclareOp

tblah committed:
Bail out if no func.func parent can be found

tblah committed:
Update tests for new naming

This is fallout from getting argument names from fir.declare operations
instead of fir.bindc attributes.

tblah commented:
Force push to include fixup commits from other PRs.

tblah committed:
Add argument to prevent per-function tbaa trees

tblah commented:
Merged manually in https://github.com/llvm/llvm-project/commit/6042c2eb9edc5c5f743c209b37a72dca664864cb
